horovod/spark/common/util.py

```diff
@@ -539,6 +543,46 @@ def _train_val_split(df, validation):
     return train_df, val_df, validation_ratio
 
 
+_FILE_AVAILABILITY_WAIT_TIMEOUT_SECS = \
+    int(os.environ.get('FILE_AVAILABILITY_WAIT_TIMEOUT_SECS', '30'))
+
+
+def _wait_file_available(store, url_list):
+    """Waiting about _FILE_AVAILABILITY_WAIT_TIMEOUT_SECS seconds (default 30 seconds) to make sure
+    all files are available for reading. This is useful in some filesystems, such as S3 which only
+    providing eventually consistency.
+    """
+    # Import LocalStore here to avoid circular import
+    from horovod.spark.common.store import LocalStore
+    if isinstance(store, LocalStore):
+        return
+
+    def wait_for_file(path):
+        end_time = time.time() + _FILE_AVAILABILITY_WAIT_TIMEOUT_SECS
+        while time.time() < end_time:
+            if store.exists(path):
+                return True
+            time.sleep(0.1)
+        return False
+
+    pool = ThreadPool(min(len(url_list), 64))
+    try:
+        results = pool.map(wait_for_file, url_list)
+        failed_list = [url for url, result in zip(url_list, results) if not result]
+        if failed_list:
+            raise RuntimeError('Timeout while waiting for all parquet-store files to appear at urls {failed_list},'
+                               'Please check whether these files were saved successfully when materializing dataframe.'
+                               .format(failed_list=','.join(failed_list)))
+    finally:
+        pool.close()
+        pool.join()
+
+
+def get_spark_df_saved_file_list(saved_path):
+    spark_session = SparkSession.builder.getOrCreate()
+    return list(spark_session.read.parquet(saved_path)._jdf.inputFiles())
```

Comment:
Why would Spark be able to see the files in `saved_path` while `store.exists(path)` returns `False`?

Reply:
Can it happen that at one point Spark sees fewer files then there have been written earlier? Like before all become synced?

Reply:
In apache/spark, this is an issue, it cannot guarantee read the full list of files.
In databricks runtime, when saving spark dataframe, an additional file `_committed` was written into the output directory, the `_committed` file includes the full list of the saved parquet files.

So this fix only guarantee correctness under databricks runtime. In petastorm spark dataset converter we use the same approach.

so do we need only enable this patch under databricks runtime ?

Reply:
And `spark_session.read.parquet(saved_path)._jdf.inputFiles()` returns the file list from that `_committed` file (in Databricks runtime)? How do you guarantee that the `_committed` file is visible to Spark at this point?

Reply:
> And spark_session.read.parquet(saved_path)._jdf.inputFiles() returns the file list from that _committed file (in Databricks runtime)?

Yes.

> How do you guarantee that the _committed file is visible to Spark at this point?

Databricks runtime code will guarantee this. It will wait the file visible and then read it.

Reply:
Makes sense to do all this effort only in Databricks runtime.